docs/contracts/function-modifiers.rst

```diff
@@ -126,13 +126,16 @@ control flow continues after the ``_`` in the preceding modifier.
     In an earlier version of Solidity, ``return`` statements in functions
     having modifiers behaved differently.
 
+.. warning::
+    Calling ``return`` directly from an assembly block in a function will return directly and skip any remaining execution in the modifier.
```

Comment:
I think that this is misleading. It sounds as if the `return` built-in in inline assembly was a legit way to return from a function with a modifier. It's a very low-level thing that should not be used lightly and I don't think it's a good idea to present it here in that way. 

Also, this is not specific to modifiers. It's just that `RETURN` is an opcode that terminates the EVM transaction. The function does not return at all. It just abruptly ends, without a chance to execute any other code. Not only modifiers will not run, but all the other high-level mechanisms provided by the compiler are bypassed too. For example if you're in an external function with a named return variable and you have assigned something to that variable, it's **not** what the function will return.

I think this is too low-level to mention here. We could have a warning that in inline assembly `return` has a completely different meaning than a high-level function `return`, but then we'd have to do that in a lot of other places. I don't think this one is special.

Reply:
I see your point, and agree that it is specific instance of a more global concept.

The reason I think the warning on modifiers is specifically important is that modifiers already have "magic" behavior to continue execution after a `return` that I don't think most devs have a clear mental model for. It would be easy to assume the modifier should always execute, and not put the pieces together that an assembly block in the function would change the modifier's behavior.

On the other hand, I wrote a script to check if there are any live contracts vulnerable to this, and only found a few on Rinkeby, so it seems this hasn't caused any problems so far. 

If you think catching this is too low-level and falls into the bucket of "things you should know before dropping into assembly", that's fair and feel free to scrap the PR.

Reply:
I would not scrap the whole PR - the note about return value when you include the placeholder multiple times is a good change and thanks for clarifying that.

I'd only remove this bit about assembly `return`. I agree that modifiers may seem magical to some devs, but this is really the problem of understanding how EVM works and I really doubt this remark will help all that much. The fact that it bypasses the high-level stuff makes `return` pretty uncommon in general, unless you're writing something inherently low-level, like a proxy.

Reply:
Alright, removed that warning, feel free to merge.